### ISSUE TITLE
PyDriller: Python Framework for Mining Software Repositories

### ISSUE THREAD

**Issue by hsanchez** [hsanchez/paper-memory#2]
## TL;DR

This paper presents Pydriller, a Python Framework that eases the process of mining Git repositories.

### Paper Link

1. [dl.acm.org/doi/10.1145/3236024.3264598](https://dl.acm.org/doi/10.1145/3236024.3264598)
2. [zora.uzh.ch/id/eprint/197754/1/fse2018td](https://www.zora.uzh.ch/id/eprint/197754/1/fse2018td.pdf)

### Author/Institution

- Davide Spadini (Delft University of Technology, Netherlands)
- Maurício Aniche (Delft University of Technology, Netherlands)
- Alberto Bacchelli (University of Zurich, Switzerland)

## Overview

This paper asserts that the extraction of information from Git repositories is not trivial and that requires a diverse set of frameworks to do it. However, these frameworks are often difficult to use. The main reason for this difficulty is the encapsulation of all the features from Git. In part, this is because developers are forced to write long and complex implementations to do the extraction and abstraction of such data.

This paper addresses this difficulty by introducing PyDriller, a framework that helps developers mine software repositories. This tool provides a simple API for (1) extracting information from Git repositories, such as commits, developers, modifications, diffs, and source code, and (2) exporting the extracted data to CSV files and or databases.

The paper compares the proposed tools with the state-of-the-art Python framework GitPython, in terms of implementation complexity, performance, and memory consumption. The results show that PyDriller requires significantly fewer lines of code to perform the same task with only a small drop in performance. A qualitative evaluation of the tool was also performed.

## Contributions and Distinctions from Previous Works

**Contributions:**

- PyDriller tool.
- Quantitative evaluation of the tool against GitPython.
- Qualitative evaluation of the tool.

**Difference:**

The most significant difference between the two tools is that GitPython offers almost all the features of Git, while PyDriller offers only features that are important when performing MSR tasks, thus hiding the underlying complexity to the end-user.

## Methods

- Wrapper around GitPython that eases the extraction of information from Git repositories.

## Results

Evaluation of tool using different metrics: lines of code (LOC) and complexity (McCabe complexity), as well as their memory consumption, and execution time.

- Execution time: PyDriller is generally slower than GitPython. This decrease in speed is expected, given that PyDriller is built on top of GitPython. However, this difference is small.
- Memory consumption: The tools behave similarly.
- LOC and complexity of the implementation: Using PyDriller results (on average) in writing 50% fewer lines of code than using GitPython. On average, the code for Pydriller is 60% less complex.
